src/diagram/Timeline.ts

```diff
@@ -0,0 +1,92 @@
+import { Cancellable } from "../command/Commands";
+import { Diagram, Memento } from "./Diagram";
+
+export class Snapshot<T extends Cancellable> {
+  constructor(readonly origin: Memento, readonly modifier: T) {}
+}
+
+export class Timeline<T extends Cancellable> {
+  private diagram!: Diagram | null;
+  private latest!: Memento;
+  private undoStack: Snapshot<T>[] = [];
+  private redoStack: T[] = [];
+
+  constructor(diagram: Diagram | null) {
```

Comment:
how about the constructor accepts null?

Reply:
https://github.com/Jerrylum/protocol-diagram/blob/562e390be789c76175c498a48f7b4eeb2e144980/src/core/Root.ts#L24C17-L24C17

It is because the child class MainDiagramHandler constructor use a super(null)

Reply:
Then you should write `super(null as any)` in class `MainDiagramHandler`. TimeLine should never accepts null

Reply:
That was good idea, updated